src/sim/pseudo_inst.cc

```diff
@@ -368,10 +368,10 @@ m5checkpoint(ThreadContext *tc, Tick delay, Tick period)
 }
 
 uint64_t
-readfile(ThreadContext *tc, Addr vaddr, uint64_t len, uint64_t offset)
+readfile(ThreadContext *tc, GuestAddr vaddr, uint64_t len, uint64_t offset)
```

Comment:
I feel like the we can simple change `Addr` to `VPtr<>`

```suggestion
readfile(ThreadContext *tc, VPtr<> vaddr, uint64_t len, uint64_t offset)
```

Details of `VPtr<>`:

The `VPtr<>` is same as `ProxyPtr<void, SETranslatingPortProxy>`
https://github.com/gem5/gem5/blob/f35815cd48a0716452a88594563c2c5131d69424/src/sim/proxy_ptr.hh#L400

`ProxyPtr<T, Proxy>` declaration
https://github.com/gem5/gem5/blob/f35815cd48a0716452a88594563c2c5131d69424/src/sim/proxy_ptr.hh#L237-L327

get arguments fro `ProxyPtr`, use `ABI::UIntPtr` type. Usually the type of `UIntPtr` is `uint32_t` if the syscall ABI is 32bits, otherwise the size of `UIntPtr` is `uint64_t`

https://github.com/gem5/gem5/blob/f35815cd48a0716452a88594563c2c5131d69424/src/sim/proxy_ptr.hh#L363-L372

Reply:
Hi @rogerchang23424, using VPtr would be the best solution, but as I mentioned in one of my earlier comments it assumes a SETranslatingPortProxy which is SE specific and makes it not usable in FS mode. And we want to be able to call m5ops in FS mode

Reply:
@giactra Sorry for missing the comments above. Currently we can't pass the github CI because of the SegmentFault of the binary `rv32uamt-ps-amoxxx_w` in SE mode. I need to find the way to solve it.

Thanks!

Reply:
Hi @rogerchang23424,
I will take a look on it tomorrow.